### PR TITLE
feat(katana): monitoring/profiling - Prometheus monitoring endpoints 

### LINKED ISSUE #1369: tracking: katana metrics
Add metrics throughout katana

```[tasklist]
### Tasks
- [ ] https://github.com/dojoengine/dojo/pull/1719
- [ ] https://github.com/dojoengine/dojo/pull/1726
- [ ] https://github.com/dojoengine/dojo/pull/1728
- [ ] https://github.com/dojoengine/dojo/pull/1791
- [ ] https://github.com/dojoengine/dojo/pull/1818
```

### LINKED ISSUE #1399: [TORII] Server - prometheus metric apis

### PR DESCRIPTION
**Pull request by hel-kame**
add: 
- `ReportMetrics` trait as template
- example of metrics integration using `pool_inserted_transactions` in `PoolMetrics` struct, part of core metrics, you can try it with the [hello dojo](https://book.dojoengine.org/cairo/hello-dojo) example with the `--metrics` flag and curl it or access it from your browser with the socket address

todo:
- readjust the `[Metrics]` derive macro with the new structure using atomic types
- writing documentation to explain the design and guide the use of metrics
- add grafana dashboard
- more metrics

Closes DOJ-134
Closed #1399
Closes #1369

### CODE REVIEW

**Review by kariy — COMMENTED** [review]
First of all, thanks for the contribution! 

Are you planning on addressing the todos in this PR ? I know this is still a draft but let's try to keep the PR as simple as possible focusing only on the basis for the metric stuff. We can address the todos in follow up PRs if necessary.

**Comment by hel-kame** [comment]
> First of all, thanks for the contribution!
> 
> Are you planning on addressing the todos in this PR ? I know this is still a draft but let's try to keep the PR as simple as possible focusing only on the basis for the metric stuff. We can address the todos in follow up PRs if necessary.

initially yes, but let's keep it simple!

File: crates/metrics/src/core_metrics.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+//! Core pool metrics.
+// This is an example of metrics implementation.
+
+use std::sync::atomic::{AtomicU64, Ordering};
+use metrics::{absolute_counter, Label};
+use crate::report_metrics::ReportMetrics;
+
+pub struct PoolMetrics {
```
Comment (kariy) [inline crates/metrics/src/core_metrics.rs:8]: The tx pool isn't being implemented properly right now so I don't think we can extract any useful metrics out of it. Fwiw, tx pool metric shouldn't be in the main `metrics` crate anyway as it's a katana thing only. Let's remove this for now.

**Comment by kariy** [comment]
I think having just the trait and derive macro should be enough for this PR. 

Personally, I would like to have metrics for RPC server connections as the first implementation  as it's a data that is fairly easy to reason about, rather than blockchain specific metrics.

Let me know if you have questions 🙂

**Comment by kariy** [comment]
Hey @hel-kame, any updates on this? I would have more capacity to work on this starting next week. So, if you can't proceed with it, I can take over PR.

**Comment by hel-kame** [comment]
Hey, I'm still working on it. I'll push some of my commits soon, at least for the RPC metrics part it's good.

**Comment by kariy** [comment]
> Are you planning on addressing the todos in this PR ? I know this is still a draft but let's try to keep the PR as simple as possible focusing only on the basis for the metric stuff. We can address the todos in follow up PRs if necessary.

> I think having just the trait and derive macro should be enough for this PR.

As mentioned, I'd very much prefer to keep this PR simple. I think setting up the traits & macros is fairly adequate for now. We can reserve the actual metrics implementation in subsequent PRs.

Let me know what you think.

**Comment by hel-kame** [comment]
Let's do that.

**Comment by glihm** [comment]
@hel-kame thanks for the time and effort here to initiate this work, @kariy has taken over this part. Could you please reach out to me on the Dojo discord? Thank you a lot man!

**Pull request closed without merging**